src/main/scala/org/constellation/p2p/Cluster.scala

```diff
@@ -0,0 +1,425 @@
+import org.constellation.p2p.PeerState.PeerState
+import org.constellation.primitives.Schema.NodeState.NodeState
+import org.constellation.primitives.Schema.{Id, NodeState}
+import org.constellation.primitives.concurrency.SingleRef
+import org.constellation.primitives.{IPManager, Schema}
+import org.constellation.util._
+import org.constellation.{ConstellationExecutionContext, DAO, PeerMetadata}
+
+import scala.concurrent.duration._
+import scala.util.Random
+
+case class SetNodeStatus(id: Id, nodeStatus: NodeState)
+
+case class PeerData(
+  peerMetadata: PeerMetadata,
+  client: APIClient,
+  notification: Seq[PeerNotification] = Seq.empty
+)
+
+case class PeerHealthCheck(status: Map[Id, Boolean])
+
+case class PendingRegistration(ip: String, request: PeerRegistrationRequest)
+
+case class Deregistration(ip: String, port: Int, id: Id)
+
+case object GetPeerInfo
+
+case class UpdatePeerInfo(peerData: PeerData)
+case class ChangePeerState(id: Id, state: NodeState)
+
+case class PeerNotification(id: Id, state: PeerState, timestamp: LocalDateTime = LocalDateTime.now()) extends Signable
+
+object PeerState extends Enumeration {
+  type PeerState = Value
+  val Leave, Join = Value
+}
+case class UpdatePeerNotifications(notifications: Seq[PeerNotification])
+
+class Cluster[F[_]: Concurrent: Logger: Timer: ContextShift](ipManager: IPManager, dao: DAO)(
+  implicit C: ContextShift[F]
+) {
+  private val nodeState: SingleRef[F, NodeState] = SingleRef[F, NodeState](NodeState.PendingDownload)
+  private val peers: SingleRef[F, Map[Id, PeerData]] = SingleRef[F, Map[Id, PeerData]](Map.empty)
+
+  def isNodeReady: F[Boolean] = nodeState.get.map(_ == NodeState.Ready)
+
+  def getPeerInfo: F[Map[Id, PeerData]] = peers.get
+
+  def updatePeerNotifications(notifications: List[PeerNotification]): F[Unit] =
+    for {
+      p <- peers.get
+      peerUpdate = notifications.flatMap { n =>
+        p.get(n.id).map { p =>
+          p.copy(notification = p.notification.diff(Seq(n)))
+        }
+      }
+      _ <- Logger[F].debug(s"Peer update: ${peerUpdate}")
+      _ <- peerUpdate.traverse(updatePeerInfo)
+    } yield ()
+
+  def removePeerRequest(hp: Option[HostPort], id: Option[Id]): F[Unit] = {
+    def filter(p: Map[Id, PeerData]) = p.filter {
+      case (pid, d) =>
+        val badHost = hp.exists {
+          case HostPort(host, port) =>
+            d.client.hostName == host && d.client.apiPort == port
+        }
+        val badId = id.contains(pid)
+        !badHost && !badId
+    }
+
+    for {
+      p <- peers.get
+      updatedPeerInfo <- peers.modify(p => (p, p).map(filter))
+
+      _ <- if (p != updatedPeerInfo) {
+        dao.metrics.incrementMetricAsync[F]("peerRemoved")
+      } else Sync[F].unit
+
+      _ <- updateMetrics()
+      _ <- updatePersistentStore()
+    } yield ()
+  }
+
+  def setNodeStatus(id: Id, state: NodeState): F[Unit] =
+    for {
+      _ <- peers.modify(
+        p =>
+          (
+            p.get(id)
```

Comment:
Maybe it could be a separate method with the name of what he does, because now it's hard to read for me it

Reply:
It is a separate method indeed (setNodeStatus)